db.go

```diff
@@ -835,34 +835,50 @@ func (db *DB) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 }
 
 // CleanTombstones re-writes any blocks with tombstones.
-func (db *DB) CleanTombstones() error {
+func (db *DB) CleanTombstones() (err error) {
 	db.cmtx.Lock()
 	defer db.cmtx.Unlock()
 
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
+	newUIDs := []ulid.ULID{}
+	defer func() {
+		// If any error is caused, we need to delete all the new directory created.
+		if err != nil {
+			for _, uid := range newUIDs {
+				dir := filepath.Join(db.Dir(), uid.String())
+				if err := os.RemoveAll(dir); err != nil {
+					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
+				}
+			}
+		}
+	}()
+
 	db.mtx.RLock()
 	blocks := db.blocks[:]
 	db.mtx.RUnlock()
 
 	deleted := []string{}
 	for _, b := range blocks {
-		ok, err := b.CleanTombstones(db.Dir(), db.compactor)
-		if err != nil {
-			return errors.Wrapf(err, "clean tombstones: %s", b.Dir())
+		ok, uid, er := b.CleanTombstones(db.Dir(), db.compactor)
+		if er != nil {
+			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
+			return err
 		}
 
 		if ok {
 			deleted = append(deleted, b.Dir())
+			newUIDs = append(newUIDs, uid)
 		}
 	}
 
 	if len(deleted) == 0 {
 		return nil
 	}
 
-	return errors.Wrap(db.reload(deleted...), "reload blocks")
+	err = errors.Wrap(db.reload(deleted...), "reload blocks")
+	return err
```

Comment:
Just `return errors.Wrap(db.reload(deleted...), "reload blocks")` is sufficient.

Reply:
I think this is because the defer block checks `err!=nil`

Reply:
Ohh, I see – that makes perfec. I think Go's semantics in this area are quite non-obvious and confusing at times. https://play.golang.org/p/RqSRG3vL4CU shows a few examples.

What I end up doing most of the time is naming the return argument. Then I can return whatever I want and defer will always bind to the actually returned value.
Unless you change scope, overwrite the named parameter with a new definition and call defer on that one (`f5`), which is quite unlikely in realistic control flow.
That seemed to be the most reliable variant for me so far.

Reply:
As I see the example in the link (`f3`), a direct return will work. I will change then.

Reply:
Yes, but only if the return param becomes named :)

Reply:
`f3()` looks most logical and IIUC this is the one you suggest as well.

`f5()` is really awkward , took few head scratches to digest :stuck_out_tongue_winking_eye: